include/swift/Basic/ProfileCounter.h

```diff
@@ -0,0 +1,42 @@
+//===------------ ProfileCounter.h - PGO Propfile counter -------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+/// \file Declares ProfileCounter, a convenient type for PGO
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_BASIC_PROFILECOUNTER_H
+#define SWIFT_BASIC_PROFILECOUNTER_H
+
+#include <cassert>
+#include <cstdint>
+
+namespace swift {
+/// A class designed to be smaller than using Optional<uint64_t> for PGO
+class ProfileCounter {
+private:
+  uint64_t count;
+
+public:
+  explicit ProfileCounter() : count(UINT64_MAX) {}
+  ProfileCounter(uint64_t count) : count(count) {}
```

Comment:
You should add an assert(hasValue()) in the constructor. And you have to check at each place where you construct such a counter if it's not UINT64_MAX.
Alternatively, if 'count' happens to be UINT64_MAX you could set it to UINT64_MAX - 1.

Reply:
I went with the 2nd option: if 'count' happens to be UINT64_MAX - set it to UINT64_MAX - 1